File: BW/10월3주차/괄호회전하기.js
/* 올바른 괄호 찾기 */
function isValid(s){
    var stack = [];
    for(var i=0; i<s.length; i++){
        var char = s[i];
        if(char === '(' || char === '[' || char === '{'){
            stack.push(char);
        }else{
            if(stack.length === 0) return false;
            var top = stack.pop();
            if((char === ')' && top != '(') ||
               (char === ']' && top != '[') ||
               (char === '}' && top != '{')){
                return false;
            }
        }
    }
    
    return stack.length === 0;
}

function solution(s) {
    var arr = s;
    var circle = [];
    var cnt = 0;
    
    /* 회전 */
    for (var i = 0; i < s.length; i++) {
        // 왼쪽으로 i칸만큼 회전 -> (i까지 자른 배열 + 0~i까지 배열)
        var rotated = s.slice(i) + s.slice(0, i);
        
        if (isValid(rotated)) {
            cnt++;
        }
    }
    
    return cnt;
}

File: MK/10월 3주차/괄호회전하기.java
import java.util.*;

class Solution {
    public int solution(String s) {
        int res=0;
        int cnt=s.length();


        while (cnt!=0) {
            cnt--;
            s = turn(s);
            res = check(s) ? res+1:res;
        }

        return res;
    }

    public String turn(String s) {
        StringBuilder sb = new StringBuilder(s);
        char c = sb.charAt(0);
        sb.deleteCharAt(0).append(c);
        return sb.toString();
    }


    public boolean check(String s) {
        Map<Character, Character> map = new HashMap<>();
        map.put(')', '(');
        map.put('}', '{');
        map.put(']', '[');

        Stack<Character> stack = new Stack<>();

        String a = "([{";

        for (Character c : s.toCharArray()) {
            if (a.contains(String.valueOf(c))) {
                stack.push(c);
            } else if (stack.size() == 0 && !a.contains(String.valueOf(c))) {
                return false;
            } else if (stack.size() !=0 && !a.contains(String.valueOf(c))) {
                if (stack.get(stack.size()-1) == map.get(c)) {
                    stack.pop();
                } else {
                    return false;
                }
            }
        }
        return stack.size() == 0;

    }
}